### PR TITLE
Create Post in List modal

### LINKED ISSUE #322: Create post in List View
Add creating post functionality in the post list view. Just like you can create a post in the buckets view.

### PR DESCRIPTION
**Pull request by LunarFang416**
## Details

- Users can now create posts in the list modal. The posts will not be sent to the canvas, and will only remain in the list modal. These posts can be downloaded to the canvas later on
- Downloading posts hasn't been implemented in the develop branch, but I found this issue a while ago and addressed it in the Delete bucket PR #305. So you wont be able to test downloading the posts from the list modal in this branch, but when #305 is merged to develop, you will be able to download posts from the list modal and it will work as we want it to

Closes #322

### CODE REVIEW

File: frontend/src/app/components/add-post-modal/add-post.component.ts [hunk 1]
```diff
@@ -123,14 +123,32 @@ export class AddPostComponent {
...
+    return await this.canvasService.createBucketPost(bucketID, post);
+  }
+
+  async addListPost() {
+    const post: Post = {
+      postID: generateUniqueID(),
+      userID: this.user.userID,
+      boardID: this.board.boardID,
+      author: this.user.username,
+      type: PostType.BUCKET,
```
Comment (Ammar-T) [inline frontend/src/app/components/add-post-modal/add-post.component.ts:135]: For list only posts, I think it'd be better to have their own type i.e. `PostType.LIST`.

File: frontend/src/app/components/add-post-modal/add-post.component.ts [hunk 2]
```diff
@@ -123,14 +123,32 @@ export class AddPostComponent {
...
+
+    return await this.canvasService.createListPost(post);
   }
 
   async handleDialogSubmit() {
     let post: Post;
 
     if (this.data.type == PostType.BUCKET && this.data.bucket) {
       post = await this.addBucketPost();
+    } else if (this.data.type == PostType.BUCKET) {
```
Comment (Ammar-T) [inline frontend/src/app/components/add-post-modal/add-post.component.ts:150]:
```suggestion
    } else if (this.data.type == PostType.LIST) {
```

**Review by JoelWiebe — COMMENTED** [review]
Thanks for adding this @LunarFang416 

Tested:
- Create post from list view containing title, message, and tags

Suggestions/ideas:
1. Post was successfully created, but I did not see any trace logs for post creation (including changes to title and message) and tags added. Can trace logs be added (or would this be included in work done in another PR for buckets already)?

Questions:
- What is PostType used for?

**Comment by LunarFang416** [comment]
@JoelWiebe I'll look into the log errors.
The PostType is basically the types of posts you can create: BOARD, BUCKETS, or LIST (new). It helps organize which posts to view in the canvas and other modals.

**Comment by LunarFang416** [comment]
@JoelWiebe Just made a few tweaks to the code. What i tested:
- POST_CREATE event is logged for post created in list modal
- POST_UPDATE logged for updated to post
- POST_TAG_ADD for tags added AFTER post created
For POST_TAG_ADD for tags added during post creation, @zarifmahmud is working on that in #315 .

Please let me know if the logs are working as expected for you as well.

**Comment by JoelWiebe** [comment]
@LunarFang416 Great work!

The following work when testing:
- POST_CREATE
- POST_DELETE
- POST_TAG_ADD
- POST_UPVOTE_ADD
- POST_UPVOTE_REMOVE
- POST_READ

However, I didn't see logs for:
- POST_TAG_REMOVE
- POST_COMMENT_ADD
- POST_COMMENT_REMOVE (is this event even implemented -- it may have been forgotten when we added the delete comment feature as this doesn't work from the canvas either -- maybe we can add this for the canvas, buckets, and list view?)

**Comment by LunarFang416** [comment]
I saw that there was an error in some old code which was causing errors in the Comment and tag removal trace logs. Now I tested POST_TAG_REMOVE, POST_COMMENT_ADD, POST_COMMENT_REMOVE. They work for the canvas, list modal, and bucket modal.

Please let me know if it works on your end as well.

File: frontend/src/app/services/canvas.service.ts [hunk 3]
```diff
@@ -56,6 +56,19 @@ export class CanvasService {
     return savedPost;
   }
 
+  async createListPost(post: Post) {
+    const savedPost = await this.postService.create(post);
+    this.socketService.emit(SocketEvent.POST_CREATE, savedPost);
```
Comment (Ammar-T) [inline frontend/src/app/services/canvas.service.ts:61]: If we emit this event, won't it show the post on other users' canvas? But, it should stay only inside the list modal.

**Comment by JoelWiebe** [comment]
@LunarFang416 @Ammar-T What do you think about us setting up a unit test suite for testing the logging of adding, moving, deleting, etc.? I just want to make sure we have everything 100% if we are using this for research analysis and publication. 

I noticed that postModifiedTitle has no value on BUCKET_ADD_POST and realize that this is a lot of manual labour to test everything (for boards, list modal, and bucket modal). It would probably be good practice and clear contract for consistent implementation if we could define and automated tests. Is this feasible/sensible to use automated testing for our trace logs given the current implementation?

**Comment by Ammar-T** [comment]
@JoelWiebe I asked Zarif about setting up a test suite and he was on board to do it. Although, haven't heard from him since so I hope that didn't scare him away haha. 

If we don't hear back within a week, we can discuss if anyone has interest in getting that started, else I could do it too.

**Comment by JoelWiebe** [comment]
@LunarFang416 I noticed a small glitch:
- When adding a post to the list view only, it does not show up immediately in the list view
- When I tagged a post in the list view, the new post that wasn't showing up did show up, but all the posts in the list view were also duplicated in the list view modal

Always so tricky, haha

**Comment by JoelWiebe** [comment]
> @JoelWiebe I asked Zarif about setting up a test suite and he was on board to do it. Although, haven't heard from him since so I hope that didn't scare him away haha.
> 
> If we don't hear back within a week, we can discuss if anyone has interest in getting that started, else I could do it too.

Thanks for asking him about this. If he's not available to work on this I'll do a really thorough manual check of everything before data collection begins. I've got other plans if you're looking for more fun features to work on!

**Comment by Ammar-T** [comment]
> Thanks for asking him about this. If he's not available to work on this I'll do a really thorough manual check of everything before data collection begins. I've got other plans if you're looking for more fun features to work on!

Sure, I can take on something new, although there's still a couple things to be addressed in this https://github.com/encorelab/ck-board/pull/307 PR. (I left some questions, you may have missed them). 

I think it'd be better to get this out of the way first, or I could work on both things concurrently - upto you.

**Comment by LunarFang416** [comment]
@Ammar-T @JoelWiebe just made a few changes to fix the issues. Hopefully they don't occur again.

File: frontend/src/app/components/list-modal/list-modal.component.ts [hunk 4]
```diff
@@ -25,6 +25,7 @@ import {
   AddPostDialog,
 } from 'src/app/components/add-post-modal/add-post.component';
 import User from 'src/app/models/user';
```
Comment (JoelWiebe) [inline frontend/src/app/components/list-modal/list-modal.component.ts:27]: Duplicate?

Comment (JoelWiebe) [reply frontend/src/app/components/list-modal/list-modal.component.ts:27]: I'm getting a few duplicate errors (i.e., for Post, PostType, and User) when the building frontend

Comment (LunarFang416) [reply frontend/src/app/components/list-modal/list-modal.component.ts:25]: I think those occurred when i merged the develop branch. Fixed now.

Comment (LunarFang416) [reply frontend/src/app/components/list-modal/list-modal.component.ts:25]: Hopefully it works properly now.

**Review by JoelWiebe — APPROVED** [review]

**Pull request merged**